src/event/sections/threshold.rs

```diff
@@ -0,0 +1,248 @@
+            WeightedThreshold::Multi(clauses) => clauses.enough_signatures(sigs),
+        }
+    }
+    pub fn to_limen(&self) -> String {
+        match self {
+            WeightedThreshold::Single(clause) => clause.to_limen(),
+            WeightedThreshold::Multi(clauses) => clauses.to_limen(),
+        }
+    }
+}
+
+impl SignatureThreshold {
+    pub fn simple(t: u64) -> Self {
+        Self::Simple(t)
+    }
+
+    pub fn single_weighted(fracs: Vec<(u64, u64)>) -> Self {
+        Self::Weighted(WeightedThreshold::Single(ThresholdClause::new_from_tuples(
+            fracs,
+        )))
+    }
+
+    pub fn multi_weighted(fracs: Vec<Vec<(u64, u64)>>) -> Self {
+        Self::Weighted(WeightedThreshold::Multi(MultiClauses::new_from_tuples(
+            fracs,
+        )))
+    }
+
+    pub fn enough_signatures(&self, sigs: &[AttachedSignaturePrefix]) -> Result<bool, Error> {
+        match self {
+            SignatureThreshold::Simple(ref t) => Ok((sigs.len() as u64) >= t.to_owned()),
+            SignatureThreshold::Weighted(ref thresh) => thresh.enough_signatures(sigs), 
+        }
+    }
+}
+
+impl Default for SignatureThreshold {
+    fn default() -> Self {
+        Self::Simple(0)
+    }
+}
+#[derive(Serialize, Deserialize, Debug, Clone, PartialEq)]
+pub struct ThresholdClause(Vec<ThresholdFraction>);
+
+impl ThresholdClause {
+    pub fn new(fracs: &[ThresholdFraction]) -> Self {
+        Self(fracs.to_owned())
+    }
+
+    pub fn new_from_tuples(tuples: Vec<(u64, u64)>) -> Self {
+        let clause = tuples
+            .into_iter()
+            .map(|(n, d)| ThresholdFraction::new(n, d))
+            .collect();
+        Self(clause)
+    }
+
+    pub fn enough_signatures(
+        &self,
+        start_index: u64,
+        sigs: &[AttachedSignaturePrefix],
+    ) -> Result<bool, Error> {
+        Ok(sigs.into_iter().fold(Zero::zero(), |acc: Fraction, sig| {
+            acc + self.0[(sig.index as u64 - start_index) as usize].fraction
+        }) >= One::one())
+    }
+
+    pub fn to_limen(&self) -> String {
+        self.0
+            .iter()
+            .map(|fr| fr.to_string())
+            .collect::<Vec<_>>()
+            .join(",")
+    }
+}
+
+#[derive(Deserialize, Serialize, Debug, Clone, PartialEq)]
+pub struct MultiClauses(Vec<ThresholdClause>);
+
+impl MultiClauses {
+    pub fn new(fracs: Vec<Vec<ThresholdFraction>>) -> Self {
+        let clauses = fracs
+            .iter()
+            .map(|clause| ThresholdClause::new(clause))
+            .collect();
+
+        Self(clauses)
+    }
+
+    pub fn new_from_tuples(fracs: Vec<Vec<(u64, u64)>>) -> Self {
+        let wt = fracs
+            .into_iter()
+            .map(|clause| ThresholdClause::new_from_tuples(clause))
+            .collect();
+        MultiClauses(wt)
+    }
+
+    pub fn enough_signatures(&self, sigs: &[AttachedSignaturePrefix]) -> Result<bool, Error> {
+        let mut out = true;
+        let mut start_index = 0u16;
```

Comment:
why do we need index calculation if we have nested vecs? perhaps my understanding is out of date

Reply:
That's because we need to keep track of which one of nested vectors contains a threshold fraction corresponding to a signature index. And signatures indexes don't match indexes of nested vectors, so I added those index calculations. Now I rewrote it to use fold instead. Hopefully, it looks a little better :D